# HG changeset patch
# User John W. Eaton <jwe@octave.org>
# Date 1463166002 14400
#      Fri May 13 15:00:02 2016 -0400
# Branch stable
# Node ID 2c6d1ef071d1104ddcc46ce403adc4b919bf9619
# Parent  44f7664689f2ce10d5444768a743a6a3961e3497
avoid crash in audiowrite argument processing (bug #47875)

* audioread.cc (Faudiowrite): Avoid out of bounds array access.

diff --git a/libinterp/dldfcn/audioread.cc b/libinterp/dldfcn/audioread.cc
--- a/libinterp/dldfcn/audioread.cc
+++ b/libinterp/dldfcn/audioread.cc
@@ -379,20 +379,26 @@ Comment.\n\
   std::string title = "";
   std::string artist = "";
   std::string comment = "";
   // Quality is currently unused?
   //
   // float quality = 0.75;
   for (int i = 3; i < nargin; i += 2)
     {
-      if (args(i).string_value () == "BitsPerSample")
+      if (i >= nargin - 1)
+        error ("audiowrite: invalid number of arguments");
+
+      std::string keyword = args(i).string_value ();
+      octave_value value_arg = args(i+1);
+
+      if (keyword == "BitsPerSample")
         {
           info.format &= ~SF_FORMAT_SUBMASK;
-          int bits = args(i + 1).int_value ();
+          int bits = value_arg.int_value ();
           if (bits == 8)
             {
               if ((info.format & SF_FORMAT_TYPEMASK) == SF_FORMAT_WAV)
                 info.format |= SF_FORMAT_PCM_U8;
               else
                 info.format |= SF_FORMAT_PCM_S8;
             }
           else if (bits == 16)
@@ -402,28 +408,28 @@ Comment.\n\
           else if (bits == 32)
             info.format |= SF_FORMAT_PCM_32;
           else
             {
               error ("audiowrite: wrong number of bits specified");
               return retval;
             }
         }
-      else if (args(i).string_value () == "BitRate")
+      else if (keyword == "BitRate")
         ;
       // Quality is currently unused?
       //
-      // else if (args(i).string_value () == "Quality")
-      //   quality = args(i + 1).int_value () * 0.01;
-      else if (args(i).string_value () == "Title")
-        title = args(i + 1).string_value ();
-      else if (args(i).string_value () == "Artist")
-        artist = args(i + 1).string_value ();
-      else if (args(i).string_value () == "Comment")
-        comment = args(i + 1).string_value ();
+      // else if (keyword == "Quality")
+      //   quality = value_arg.int_value () * 0.01;
+      else if (keyword == "Title")
+        title = value_arg.string_value ();
+      else if (keyword == "Artist")
+        artist = value_arg.string_value ();
+      else if (keyword == "Comment")
+        comment = value_arg.string_value ();
       else
         {
           error ("audiowrite: wrong argument name");
           return retval;
         }
     }
 
   SNDFILE *file = sf_open (filename.c_str (), SFM_WRITE, &info);
